File: tests/test_wally_opencode.cpp
#include "test_common.h"

#include <chrono>
#include <cstdlib>
#include <filesystem>
#include <nlohmann/json.hpp>
#include <stdexcept>
#include <string>
#include <vector>

#include "account/console.h"
#include "account/credentials.h"
#include "harness/opencode.h"

namespace {

namespace fs = std::filesystem;
using Json = nlohmann::json;

class Environment {
   public:
    Environment(const char* name, const char* value) : name_(name) {
        if (const char* previous = std::getenv(name)) {
            had_previous_ = true;
            previous_ = previous;
        }
        Set(value);
    }
    ~Environment() { Set(had_previous_ ? previous_.c_str() : nullptr); }

   private:
    void Set(const char* value) {
#if defined(_WIN32)
        _putenv_s(name_.c_str(), value != nullptr ? value : "");
#else
        value != nullptr ? setenv(name_.c_str(), value, 1) : unsetenv(name_.c_str());
#endif
    }
    std::string name_;
    std::string previous_;
    bool had_previous_ = false;
};

class TemporaryDirectory {
   public:
    TemporaryDirectory() {
        const auto nonce = std::chrono::steady_clock::now().time_since_epoch().count();
        path_ = fs::temp_directory_path() / ("wally-opencode-test-" + std::to_string(nonce));
        fs::create_directories(path_);
    }
    ~TemporaryDirectory() {
        std::error_code ignored;
        fs::remove_all(path_, ignored);
    }
    const fs::path& path() const { return path_; }

   private:
    fs::path path_;
};

long long Now() {
    return std::chrono::duration_cast<std::chrono::seconds>(
               std::chrono::system_clock::now().time_since_epoch())
        .count();
}

bool Seed(const fs::path& profile, long long expires_at, std::string* error) {
    Environment scoped_profile("WALLY_PROFILE_DIR", profile.string().c_str());
    wally::account::Credentials credentials;
    credentials.console_url = "https://console.runanywhere.ai";
    credentials.email = "developer@example.test";
    credentials.access_token = "old-access-token";
    credentials.refresh_token = "refresh-token";
    credentials.expires_at = expires_at;
    return wally::account::Save(credentials, error);
}

TestResult test_ephemeral_config_and_passthrough() {
    TestResult result;
    result.test_name = "ephemeral_config_and_passthrough";
    TemporaryDirectory temporary;
    Environment profile("WALLY_PROFILE_DIR", temporary.path().string().c_str());
    Environment existing("OPENCODE_CONFIG_CONTENT", "{\"keep\":true}");
    std::string error;
    if (!Seed(temporary.path(), Now() + 3600, &error)) {
        result.details = error;
        return result;
    }

    bool spawned = false;
    const std::vector<std::string> arguments = {"run", "--agent", "build", "two words"};
    // The launch path verifies the session against the console before it starts
    // anything, so the console has to be answered here. A default ConsoleClient
    // would put a real request on the wire, which a unit test must never do.
    const wally::account::ConsoleClient console(
        [](const wally::account::HttpRequest& request, wally::account::HttpResponse* response,
           std::string*) {
            if (!request.url.ends_with("/v1/auth/me")) {
                return false;
            }
            response->status = 200;
            response->body = Json{{"email", "developer@example.test"}}.dump();
            return true;
        });
    const int status = wally::harness::LaunchOpenCodeCloud(
        "glm-5.3", arguments, console,
        [&](const std::string& executable, const std::vector<std::string>& received) {
            spawned = true;
            if (executable != "opencode" || received != arguments) {
                return 91;
            }
            const char* raw = std::getenv("OPENCODE_CONFIG_CONTENT");
            if (raw == nullptr) {
                return 92;
            }
            const Json config = Json::parse(raw);
            const Json& provider = config.at("provider").at("runanywhere");
            if (config.at("model") != "runanywhere/glm-5.3" ||
                provider.at("npm") != "@ai-sdk/openai-compatible" ||
                provider.at("options").at("baseURL") != "https://console.runanywhere.ai/v1" ||
                provider.at("options").at("apiKey") != "old-access-token" ||
                provider.at("models").at("glm-5.3").at("name") != "glm-5.3") {
                return 93;
            }
            return 0;
        });
    const char* restored = std::getenv("OPENCODE_CONFIG_CONTENT");
    if (status != 0 || !spawned || restored == nullptr ||
        std::string(restored) != "{\"keep\":true}") {
        result.details = "launch did not preserve arguments and the existing environment";
        result.actual = std::to_string(status);
        return result;
    }
    if (std::distance(fs::directory_iterator(temporary.path()), fs::directory_iterator{}) != 1) {
        result.details = "launch wrote a tool or project configuration file";
        return result;
    }
    result.passed = true;
    return result;
}

TestResult test_refreshes_expired_session_without_sdk_bootstrap() {
    TestResult result;
    result.test_name = "refreshes_expired_session_without_sdk_bootstrap";
    TemporaryDirectory temporary;
    Environment profile("WALLY_PROFILE_DIR", temporary.path().string().c_str());
    Environment api_key("RUNANYWHERE_API_KEY", nullptr);
    Environment environment("RUNANYWHERE_ENVIRONMENT", nullptr);
    Environment config("OPENCODE_CONFIG_CONTENT", nullptr);
    std::string error;
    if (!Seed(temporary.path(), Now() - 1, &error)) {
        result.details = error;
        return result;
    }

    bool refreshed = false;
    wally::account::ConsoleClient console([&](const wally::account::HttpRequest& request,
                                             wally::account::HttpResponse* response, std::string*) {
        if (request.url.ends_with("/v1/auth/me")) {
            response->status = 200;
            response->body = Json{{"email", "developer@example.test"}}.dump();
            return true;
        }
        if (!request.url.ends_with("/v1/auth/cli/refresh") || request.bearer_token.size() != 0 ||
            Json::parse(request.body).at("refresh_token") != "refresh-token") {
            return false;
        }
        refreshed = true;
        response->status = 200;
        response->body = Json{{"access_token", "new-access-token"},
                              {"refresh_token", "new-refresh-token"},
                              {"expires_in", 7200}}
                             .dump();
        return true;
    });

    const int status = wally::harness::LaunchOpenCodeCloud(
        "hosted-model", {}, console, [&](const std::string&, const std::vector<std::string>&) {
            const Json value = Json::parse(std::getenv("OPENCODE_CONFIG_CONTENT"));
            return value.at("provider").at("runanywhere").at("options").at("apiKey") ==
                           "new-access-token"
                       ? 0
                       : 94;
        });
    if (status != 0 || !refreshed || std::getenv("OPENCODE_CONFIG_CONTENT") != nullptr) {
        result.details = "expired cloud credentials were not refreshed ephemerally";
        result.actual = std::to_string(status);
        return result;
    }

    wally::account::Credentials stored;
    if (!wally::account::Load(&stored, &error) || stored.access_token != "new-access-token" ||
        stored.refresh_token != "new-refresh-token" || stored.expires_at <= Now()) {
        result.details = error.empty() ? "refreshed session was not stored" : error;
        return result;
    }
    result.passed = true;
    return result;
}

TestResult test_restores_config_when_spawn_throws() {
    TestResult result;
    result.test_name = "restores_config_when_spawn_throws";
    TemporaryDirectory temporary;
    Environment profile("WALLY_PROFILE_DIR", temporary.path().string().c_str());
    Environment config("OPENCODE_CONFIG_CONTENT", "original-value");
    std::string error;
    if (!Seed(temporary.path(), Now() + 3600, &error)) {
        result.details = error;
        return result;
    }

    bool threw = false;
    try {
        const wally::account::ConsoleClient console(
            [](const wally::account::HttpRequest& request, wally::account::HttpResponse* response,
               std::string*) {
                if (!request.url.ends_with("/v1/auth/me")) {
                    return false;
                }
                response->status = 200;
                response->body = Json{{"email", "developer@example.test"}}.dump();
                return true;
            });
        static_cast<void>(wally::harness::LaunchOpenCodeCloud(
            "hosted-model", {}, console,
            [](const std::string&, const std::vector<std::string>&) -> int {
                throw std::runtime_error("synthetic spawn failure");
            }));
    } catch (const std::runtime_error&) {
        threw = true;
    }
    const char* restored = std::getenv("OPENCODE_CONFIG_CONTENT");
    if (!threw || restored == nullptr || std::string(restored) != "original-value") {
        result.details = "temporary OpenCode config survived an exceptional child launch";
        return result;
    }
    result.passed = true;
    return result;
}

}  // namespace

TestResult test_config_injects_limit_and_cost() {
    TestResult result;
    result.test_name = "config_injects_limit_and_cost";
    using Json = nlohmann::json;

    // glm-like: 1M context, no separate output cap (0 -> sane default), and
    // $0.60/$2.20 per Mtok (600000/2200000 micro-dollars).
    const Json j = Json::parse(wally::harness::BuildOpenCodeCloudConfig(
        "glm-5.3-flash", "https://x/v1", "tok", 1048576, 0, 600000, 2200000));
    const Json m = j["provider"]["runanywhere"]["models"]["glm-5.3-flash"];
    const double cin = m.value("cost", Json::object()).value("input", -1.0);
    const double cout = m.value("cost", Json::object()).value("output", -1.0);
    if (m["limit"]["context"].get<std::int64_t>() != 1048576 ||
        m["limit"]["output"].get<std::int64_t>() != 65536 ||
        cin < 0.5999 || cin > 0.6001 || cout < 2.1999 || cout > 2.2001) {
        result.details = "limit/cost injection wrong: " + m.dump();
        return result;
    }
    // No metadata (all zeros) -> neither block is emitted.
    const Json bare = Json::parse(
        wally::harness::BuildOpenCodeCloudConfig("m", "https://x/v1", "tok", 0, 0, 0, 0));
    if (bare["provider"]["runanywhere"]["models"]["m"].contains("limit") ||
        bare["provider"]["runanywhere"]["models"]["m"].contains("cost")) {
        result.details = "empty metadata should omit limit and cost";
        return result;
    }
    result.passed = true;
    return result;
}

int main(int argc, char** argv) {
    TestSuite suite("wally_opencode");
    suite.add("config_injects_limit_and_cost", test_config_injects_limit_and_cost);
    suite.add("ephemeral_config_and_passthrough", test_ephemeral_config_and_passthrough);
    suite.add("refreshes_expired_session_without_sdk_bootstrap",
              test_refreshes_expired_session_without_sdk_bootstrap);
    suite.add("restores_config_when_spawn_throws", test_restores_config_when_spawn_throws);
    return suite.run(argc, argv);
}

File: tests/test_wally_opencode.rs
//! Port of tests/test_wally_opencode.cpp.

mod common;

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use serde_json::{json, Value};

use wally::account::{self, ConsoleClient, Credentials, HttpRequest, HttpResponse};
use wally::harness::{self, CatalogModel, SpawnFunction};

use common::{env_lock, EnvGuard};

fn now_seconds() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}

/// Writes the same fixed credential the C++ Seed() helper wrote. Callers must
/// already have WALLY_PROFILE_DIR pointed at an isolated directory (an
/// EnvGuard held for the test's whole body).
fn seed(expires_at: i64) -> Result<(), String> {
    let credentials = Credentials {
        console_url: "https://console.runanywhere.ai".to_string(),
        email: "developer@example.test".to_string(),
        access_token: "old-access-token".to_string(),
        refresh_token: "refresh-token".to_string(),
        expires_at,
    };
    account::save(&credentials)
}

// These exercise wally::account::ConsoleClient::who_am_i/refresh/fetch_models/
// fetch_catalog and wally::account::{load,save}, which remain `todo!()` as of
// this writing; ported faithfully, they will panic there until that port
// lands.

#[test]
fn ephemeral_config_and_passthrough() {
    let _lock = env_lock();
    let mut env = EnvGuard::new();
    let temporary = tempfile::tempdir().expect("temp dir");
    env.set("WALLY_PROFILE_DIR", temporary.path());
    env.set("OPENCODE_CONFIG_CONTENT", "{\"keep\":true}");
    seed(now_seconds() + 3600).expect("seed credentials");

    // The launch path verifies the session against the console before it
    // starts anything, so the console has to be answered here. A default
    // ConsoleClient would put a real request on the wire, which a unit test
    // must never do.
    let console = ConsoleClient::new(Some(Arc::new(|request: &HttpRequest| {
        if !request.url.ends_with("/v1/me") {
            return Err("unexpected request".to_string());
        }
        Ok(HttpResponse {
            status: 200,
            body: json!({ "email": "developer@example.test" }).to_string(),
            ..Default::default()
        })
    })));

    let spawned_flag = Arc::new(AtomicBool::new(false));
    let spawned = spawned_flag.clone();
    let arguments = vec![
        "run".to_string(),
        "--agent".to_string(),
        "build".to_string(),
        "two words".to_string(),
    ];
    let expected_arguments = arguments.clone();
    let spawn: SpawnFunction = Arc::new(move |executable: &str, received: &[String]| {
        spawned.store(true, Ordering::SeqCst);
        if executable != "opencode" || received != expected_arguments.as_slice() {
            return 91;
        }
        let raw = match std::env::var("OPENCODE_CONFIG_CONTENT") {
            Ok(value) => value,
            Err(_) => return 92,
        };
        let config: Value = match serde_json::from_str(&raw) {
            Ok(v) => v,
            Err(_) => return 93,
        };
        let provider = &config["provider"]["runanywhere"];
        if config["model"] != json!("runanywhere/glm-5.3")
            || provider["npm"] != json!("@ai-sdk/openai-compatible")
            || provider["options"]["baseURL"] != json!("https://console.runanywhere.ai/v1")
            || provider["options"]["apiKey"] != json!("old-access-token")
            || provider["models"]["glm-5.3"]["name"] != json!("glm-5.3")
        {
            return 93;
        }
        0
    });

    let status = harness::launch_open_code_cloud_with("glm-5.3", &arguments, &console, &spawn);
    let restored = std::env::var("OPENCODE_CONFIG_CONTENT").ok();
    assert!(
        status == 0
            && spawned_flag.load(Ordering::SeqCst)
            && restored.as_deref() == Some("{\"keep\":true}"),
        "launch did not preserve arguments and the existing environment (status={status})"
    );
    let entries = std::fs::read_dir(temporary.path())
        .expect("read profile dir")
        .count();
    assert_eq!(
        entries, 1,
        "launch wrote a tool or project configuration file"
    );
}

#[test]
fn refreshes_expired_session_without_sdk_bootstrap() {
    let _lock = env_lock();
    let mut env = EnvGuard::new();
    let temporary = tempfile::tempdir().expect("temp dir");
    env.set("WALLY_PROFILE_DIR", temporary.path());
    env.unset("RUNANYWHERE_API_KEY");
    env.unset("RUNANYWHERE_ENVIRONMENT");
    env.unset("OPENCODE_CONFIG_CONTENT");
    seed(now_seconds() - 1).expect("seed credentials");

    let refreshed_flag = Arc::new(AtomicBool::new(false));
    let refreshed = refreshed_flag.clone();
    let console = ConsoleClient::new(Some(Arc::new(move |request: &HttpRequest| {
        if request.url.ends_with("/v1/me") {
            return Ok(HttpResponse {
                status: 200,
                body: json!({ "email": "developer@example.test" }).to_string(),
                ..Default::default()
            });
        }
        let body: Value = serde_json::from_str(&request.body).unwrap_or(Value::Null);
        if !request.url.ends_with("/auth/cli/refresh")
            || !request.bearer_token.is_empty()
            || body["refresh_token"] != json!("refresh-token")
        {
            return Err("unexpected request".to_string());
        }
        refreshed.store(true, Ordering::SeqCst);
        Ok(HttpResponse {
            status: 200,
            body: json!({
                "access_token": "new-access-token",
                "refresh_token": "new-refresh-token",
                "expires_in": 7200,
            })
            .to_string(),
            ..Default::default()
        })
    })));

    let spawn: SpawnFunction = Arc::new(|_executable: &str, _received: &[String]| {
        let raw = std::env::var("OPENCODE_CONFIG_CONTENT").unwrap_or_default();
        let value: Value = serde_json::from_str(&raw).unwrap_or(Value::Null);
        if value["provider"]["runanywhere"]["options"]["apiKey"] == json!("new-access-token") {
            0
        } else {
            94
        }
    });

    let status = harness::launch_open_code_cloud_with("hosted-model", &[], &console, &spawn);
    assert!(
        status == 0
            && refreshed_flag.load(Ordering::SeqCst)
            && std::env::var("OPENCODE_CONFIG_CONTENT").is_err(),
        "expired cloud credentials were not refreshed ephemerally (status={status})"
    );

    let stored = account::load().expect("load stored credentials");
    assert_eq!(stored.access_token, "new-access-token");
    assert_eq!(stored.refresh_token, "new-refresh-token");
    assert!(
        stored.expires_at > now_seconds(),
        "refreshed session was not stored"
    );
}

#[test]
fn restores_config_when_spawn_throws() {
    let _lock = env_lock();
    let mut env = EnvGuard::new();
    let temporary = tempfile::tempdir().expect("temp dir");
    env.set("WALLY_PROFILE_DIR", temporary.path());
    env.set("OPENCODE_CONFIG_CONTENT", "original-value");
    seed(now_seconds() + 3600).expect("seed credentials");

    let console = ConsoleClient::new(Some(Arc::new(|request: &HttpRequest| {
        if !request.url.ends_with("/v1/me") {
            return Err("unexpected request".to_string());
        }
        Ok(HttpResponse {
            status: 200,
            body: json!({ "email": "developer@example.test" }).to_string(),
            ..Default::default()
        })
    })));
    let spawn: SpawnFunction = Arc::new(|_executable: &str, _received: &[String]| {
        panic!("synthetic spawn failure");
    });

    // Rust has no exceptions; a panicking `spawn` is the equivalent of the
    // C++ test's throwing lambda, and unwinding still runs ScopedOpenCodeConfig's
    // Drop, exactly as a C++ exception still ran ScopedOpenCodeConfig's destructor.
    let outcome = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
        harness::launch_open_code_cloud_with("hosted-model", &[], &console, &spawn)
    }));
    let threw = outcome.is_err();
    let restored = std::env::var("OPENCODE_CONFIG_CONTENT").ok();
    assert!(
        threw && restored.as_deref() == Some("original-value"),
        "temporary OpenCode config survived an exceptional child launch"
    );
}

#[test]
fn config_injects_limit_and_cost() {
    // glm-like: 1M context, no separate output cap (0 -> sane default), and
    // $0.60/$2.20 per Mtok (600000/2200000 micro-dollars).
    let j: Value = serde_json::from_str(&harness::build_open_code_cloud_config(
        "glm-5.3-flash",
        "https://x/v1",
        "tok",
        &[CatalogModel {
            id: "glm-5.3-flash".to_string(),
            context_window: 1048576,
            max_output: 0,
            input_per_mtok: 600000,
            output_per_mtok: 2200000,
        }],
    ))
    .expect("parse config");
    let m = &j["provider"]["runanywhere"]["models"]["glm-5.3-flash"];
    let cin = m["cost"]["input"].as_f64().unwrap_or(-1.0);
    let cout = m["cost"]["output"].as_f64().unwrap_or(-1.0);
    assert!(
        m["limit"]["context"] == json!(1048576)
            && m["limit"]["output"] == json!(65536)
            && (0.5999..=0.6001).contains(&cin)
            && (2.1999..=2.2001).contains(&cout),
        "limit/cost injection wrong: {m:?}"
    );

    // No metadata (all zeros) -> neither block is emitted.
    let bare: Value = serde_json::from_str(&harness::build_open_code_cloud_config(
        "m",
        "https://x/v1",
        "tok",
        &[CatalogModel {
            id: "m".to_string(),
            ..Default::default()
        }],
    ))
    .expect("parse config");
    let bare_model = &bare["provider"]["runanywhere"]["models"]["m"];
    assert!(
        bare_model.get("limit").is_none() && bare_model.get("cost").is_none(),
        "empty metadata should omit limit and cost"
    );

    // Every catalog model becomes a selectable entry; the launched one is default.
    let all: Value = serde_json::from_str(&harness::build_open_code_cloud_config(
        "glm-5.3-flash",
        "https://x/v1",
        "tok",
        &[
            CatalogModel {
                id: "glm-5.3-flash".to_string(),
                ..Default::default()
            },
            CatalogModel {
                id: "qwen3.8-27b".to_string(),
                ..Default::default()
            },
            CatalogModel {
                id: "gemma-4".to_string(),
                ..Default::default()
            },
        ],
    ))
    .expect("parse config");
    let all_models = all["provider"]["runanywhere"]["models"]
        .as_object()
        .expect("models object");
    assert!(
        all_models.len() == 3 && all["model"] == json!("runanywhere/glm-5.3-flash"),
        "all catalog models must appear, primary as default: {all:?}"
    );
}
